src/core/mac/mac_frame.cpp

```diff
@@ -145,6 +145,127 @@ ThreadError Frame::InitMacHeader(uint16_t aFcf, uint8_t aSecurityControl)
     return kThreadError_None;
 }
 
+ThreadError Frame::ValidatePsdu(void)
+{
+    ThreadError error = kThreadError_Parse;
+    uint8_t offset = 0;
+    uint16_t fcf;
+    uint8_t footerLength = kFcsSize;
+
+    VerifyOrExit((offset += kFcfSize + kDsnSize) <= GetPsduLength(), ;);
+    fcf = static_cast<uint16_t>((GetPsdu()[1] << 8) | GetPsdu()[0]);
+
+    // Destinatinon PAN + Address
+    switch (fcf & Frame::kFcfDstAddrMask)
+    {
+    case Frame::kFcfDstAddrNone:
+        break;
+
+    case Frame::kFcfDstAddrShort:
+        offset += sizeof(PanId) + sizeof(ShortAddress);
+        break;
+
+    case Frame::kFcfDstAddrExt:
+        offset += sizeof(PanId) + sizeof(ExtAddress);
+        break;
+
+    default:
+        goto exit;
+    }
+
+    // Source PAN + Address
+    switch (fcf & Frame::kFcfSrcAddrMask)
+    {
+    case Frame::kFcfSrcAddrNone:
+        break;
+
+    case Frame::kFcfSrcAddrShort:
+        if ((fcf & Frame::kFcfPanidCompression) == 0)
+        {
+            offset += sizeof(PanId);
+        }
+
+        offset += sizeof(ShortAddress);
+        break;
+
+    case Frame::kFcfSrcAddrExt:
+        if ((fcf & Frame::kFcfPanidCompression) == 0)
+        {
+            offset += sizeof(PanId);
+        }
+
+        offset += sizeof(ExtAddress);
+        break;
+
+    default:
+        goto exit;
+    }
+
+    VerifyOrExit(offset <= GetPsduLength(), ;);
+
+    // Security Header
+    if (fcf & Frame::kFcfSecurityEnabled)
+    {
+        VerifyOrExit((offset += kSecurityControlSize + kFrameCounterSize) <= GetPsduLength(), ;);
+        uint8_t secControl = GetPsdu()[offset];
```

Comment:
Reading the Security Control byte should happen before the offset increment in the line above.